### ISSUE TITLE
多轮对话grpo训练出现问题

### ISSUE THREAD

**Issue by warriorandroyal** [modelscope/ms-swift#7214]
训练一开始都是正常的，训练到100步就会突然开始只完成第一轮的回复，不再有第二轮的回复了，并且第一轮的输出全部都是！！！！！！
<img width="1788" height="1181" alt="Image" src="https://github.com/user-attachments/assets/75a86aeb-61dd-46fc-b42b-dbfbb5156da6" />

下面是我写的多轮对话的插件
import re
import torch
from typing import List, Dict, Union, Tuple, Optional

# =================================================================
# 1. 核心工具函数 (Utils)
# =================================================================

def check_format_strict(text: str) -> float:
    """检查是否同时包含 <think> 和 <answer>"""
    if not text: return 0.0
    has_think = re.search(r'<think>.*?</think>', text, re.DOTALL | re.IGNORECASE)
    has_answer = re.search(r'<answer>.*?</answer>', text, re.DOTALL | re.IGNORECASE)
    return 1.0 if (has_think and has_answer) else 0.0

def extract_answer_content(text: str) -> str:
    """提取 <answer> 标签内的文本"""
    if not text: return ""
    match = re.search(r'<answer>(.*?)</answer>', text, re.DOTALL | re.IGNORECASE)
    return match.group(1).strip().lower() if match else ""

def parse_ground_truth_extended(sol) -> Tuple[int, int]:
    """解析 Ground Truth: (label, agent1_pred)"""
    try:
        val = sol.item() if hasattr(sol, 'item') else sol
        val_str = str(val)
        if ',' in val_str:
            parts = val_str.split(',')
            return int(parts[0]), int(parts[1])
        return int(float(val_str)), -1
    except:
        return -100, -1

def get_advocate_content(messages: List[Dict[str, str]]) -> str:
    """
    从 messages 历史中提取 Advocate (Agent 2) 的回复。
    根据你的日志结构：
    Index 0: System
    Index 1: User (Input + Agent 1 Verdict)
    Index 2: Assistant (Advocate / Agent 2) <--- 我们要找这个
    Index 3: User (Judge Prompt)
    """
    # 防御性编程：确保列表长度足够且角色正确
    if len(messages) > 2 and messages[2].get('role') == 'assistant':
        return messages[2].get('content', "")
    return ""

# =================================================================
# 2. 调度器 (GoAScheduler) - 添加 EOS 移除逻辑
# =================================================================

class GoAScheduler(MultiTurnScheduler):
    # 这是完整的 Judge Prompt
    judge_trigger_prompt = """**SYSTEM INSTRUCTION:**
。。。。。
**FINAL OUTPUT:**
<think>
[Evidence Check: Scan raw reviews. Are the complaints actually "Section A" (Defects) or "Section B" (Exclusions)? Cite specific examples.]
[Agent Review: Does Agent 2's audit hold up against the Raw Evidence?]
[Math Check: Valid Count = ?, Total = ?, Ratio = ?]
[Threshold Verdict: Does it pass the >3 Count and >=20% Ratio rule?]
</think>
<answer>Yes/No</answer>
**Start your response immediately with <think> and end with </answer>.**"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._judge_token_ids = None

    def _get_judge_token_ids(self, tokenizer) -> List[int]:
        if self._judge_token_ids is None:
            self._judge_token_ids = tokenizer.encode(self.judge_trigger_prompt, add_special_tokens=False)
        return self._judge_token_ids

    def check_finished(self, infer_request: 'RolloutInferRequest', response_choice: 'ChatCompletionResponseChoice',
                       current_turn: int) -> bool:
        if current_turn >= 2:
            return True
        return super().check_finished(infer_request, response_choice, current_turn)

    def step(self, infer_request: 'RolloutInferRequest', response_choice: 'ChatCompletionResponseChoice',
             current_turn: int) -> Dict:
        response_token_ids = list(response_choice.token_ids) if response_choice.token_ids else []
        
        # --- 提取 Logprobs (保持原样) ---
        raw_logprobs = self._extract_logprobs_from_choice(response_choice)
        rollout_logprobs = []
        if raw_logprobs:
            for item in raw_logprobs:
                if isinstance(item, dict):
                    rollout_logprobs.append(item.get('logprob', 0.0))
                elif isinstance(item, (float, int)):
                    rollout_logprobs.append(float(item))
                elif hasattr(item, 'logprob'):
                    rollout_logprobs.append(item.logprob)
                else:
                    rollout_logprobs.append(0.0)
        # ---------------------

        tokenizer = self.tokenizer
        if tokenizer is None:
            raise ValueError("Tokenizer is missing.")

        if current_turn == 1 and hasattr(tokenizer, 'eos_token_id') and tokenizer.eos_token_id is not None:
            if response_token_ids and response_token_ids[-1] == tokenizer.eos_token_id:
                # 移除 Token ID
                response_token_ids = response_token_ids[:-1]
                # 移除对应的 Logprob，保持长度对齐
                if len(rollout_logprobs) == len(response_token_ids) + 1:
                    rollout_logprobs = rollout_logprobs[:-1]

        # 这里的 mask 对应 response_token_ids，全为 1 即可
        response_loss_mask = [1] * len(response_token_ids)

        if current_turn == 1:
            # 1. 幂等性检查
            is_prompt_exist = (
                len(infer_request.messages) > 0 and 
                self.judge_trigger_prompt in infer_request.messages[-1].get('content', '')
            )
            
            if not is_prompt_exist:
                infer_request.messages.append({'role': 'user', 'content': self.judge_trigger_prompt})

        return {
            'infer_request': infer_request,
            'response_token_ids': response_token_ids,
            'response_loss_mask': response_loss_mask,
            'rollout_logprobs': rollout_logprobs
        }

# =================================================================
# 3. 奖励函数 (Rewards) - 逻辑修正版
# =================================================================

class AdvocateFormatReward(ORM):
    def __call__(self, completions, **kwargs) -> List[float]:
        """
        计算 Advocate 的格式分。
        来源：kwargs['messages'] (历史对话)
        """
        messages_batch = kwargs.get('messages', [])
        rewards = []
        
        # 遍历 batch 中的每一条对话历史
        for messages in messages_batch:
            # 1. 从历史中提取 Advocate 的文本
            adv_text = get_advocate_content(messages)
            # 2. 计算奖励
            rewards.append(check_format_strict(adv_text))
            
        return rewards

class AdvocateAccuracyReward(ORM):
    def __call__(self, completions, ground_truth, **kwargs) -> List[float]:
        """
        计算 Advocate 的准确率分。
        来源：kwargs['messages'] (历史对话)
        """
        messages_batch = kwargs.get('messages', [])
        rewards = []
        
        for messages, sol in zip(messages_batch, ground_truth):
            gt, agent1_pred = parse_ground_truth_extended(sol)
            if gt == -100:
                rewards.append(0.0)
                continue
            
            # 1. 从历史中提取 Advocate 的文本
            adv_text = get_advocate_content(messages)
            ans = extract_answer_content(adv_text)
            
            adv_pred = -1
            if 'yes' in ans: adv_pred = 1
            elif 'no' in ans: adv_pred = 0
            
            # 2. 计算奖励逻辑 (保持你原有的逻辑不变)
            score = 0.0
            if adv_pred == gt:
                if agent1_pred == -1: score = 1.0 
                elif agent1_pred == gt: score = 1.0 
                else: score = 2.0 
            
            rewards.append(score)
        return rewards

class JudgeFormatReward(ORM):
    def __call__(self, completions, **kwargs) -> List[float]:
        """
        计算 Judge 的格式分。
        来源：completions (当前生成的文本)
        """
        rewards = []
        # 直接检查生成的 completion，因为这就是 Judge 的回复
        for content in completions:
            rewards.append(check_format_strict(content))
        return rewards

class JudgeAccuracyReward(ORM):
    def __call__(self, completions, ground_truth, **kwargs) -> List[float]:
        """
        计算 Judge 的准确率分。
        来源：completions (当前生成的文本)
        """
        rewards = []
        for content, sol in zip(completions, ground_truth):
            gt, _ = parse_ground_truth_extended(sol)
            if gt == -100:
                rewards.append(0.0)
                continue
            
            # 直接从 completion 提取答案
            ans = extract_answer_content(content)
            pred = -1
            if 'yes' in ans: pred = 1
            elif 'no' in ans: pred = 0
            
            rewards.append(1.0 if pred == gt else 0.0)
        return rewards

# 注册组件
multi_turns['goa_scheduler'] = GoAScheduler
orms['advocate_format'] = AdvocateFormatReward
orms['advocate_acc'] = AdvocateAccuracyReward
orms['judge_format'] = JudgeFormatReward
orms['judge_acc'] = JudgeAccuracyReward

下面是我的训练脚本
MASTER_PORT=$PORT0 \
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=8 \
swift rlhf \
    --rlhf_type grpo \
    --model Qwen/Qwen2___5-7B-Instruct \
    --model_type qwen2_5 \
    --output_dir "$output_dir" \
    --datasetrl_train.jsonl \
    --external_plugins examples/train/grpo/plugin/plugin.py \
    --multi_turn_scheduler goa_scheduler \
    --reward_funcs advocate_format judge_format advocate_acc judge_acc \
    --reward_weights 1 1 1 1 \
    --overlong_filter True \
    --epsilon_high 0.28 \
    --loss_type bnpo \
    --add_version False \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.5 \
    --vllm_tensor_parallel_size 4 \
    --offload_optimizer true \
    --offload_model true \
    --vllm_max_model_len 16000 \
    --sleep_level 1 \
    --train_type full \
    --torch_dtype bfloat16 \
    --max_length 16000 \
    --max_completion_length 3072 \
    --truncation_strategy left \
    --num_train_epochs 2 \
    --max_epochs 2 \
    --dynamic_sample True \
    --max_resample_times 3 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 2 \
    --eval_steps 100 \
    --save_steps 100 \
    --run_name v1\
    --project_name biyelunwen\
    --save_total_limit 3 \
    --logging_steps 1 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --deepspeed zero3 \
    --log_completions true \
    --log_entropy true \
    --report_to wandb \
    --temperature 1.0 \
    --top_p 1.0 \
    --top_k 80 \
    "${resume_arg[@]}"

**Comment by hjh0119** [comment]
输出感叹号是模型训崩了

**Comment by warriorandroyal** [comment]
> 输出感叹号是模型训崩了

对，这种是什么原因看得出来吗

**Comment by hjh0119** [comment]
检查一下有没有梯度爆炸

**Comment by zhenghuawang6** [comment]
我现在也有这个问题，大概的问题是如果同一个批次的样本有不同的轮次的话，就会出现 reward 不一致的问题。

**Comment by z0o0ey** [comment]
please retry with latest swift and vllm, feel free to reopen or create a new issue if it the problem persists.
